README.md
# v2e [![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT) [![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/drive/1czx-GJnx-UkhFVBbfoACLVZs8cYlcr_M?usp=sharing)

Python torch + opencv code to go from conventional stroboscopic video frames with low frame rate into realistic synthetic DVS event streams with much higher effective timing precision. v2e includes finite intensity-depenedent photoreceptor bandwidth, Gaussian pixel to pixel event threshold variation, and noise 'leak' events.

See the [v2e home page](https://sites.google.com/view/video2events/home) for videos and further information. 

Our paper (below) about _v2e_ **debunks incorrect assertions about event cameras that pervade the current computer vision literature.**

See [v2e-sample-input-data](https://drive.google.com/drive/folders/1oWxTB9sMPp6UylAdxg5O1ko_GaIXz7wo?usp=sharing) for sample input files.

![v2e-tennis-example](media/v2e-tennis-split-screen.gif)

Vote for [new v2e features](https://docs.google.com/forms/d/e/1FAIpQLSdJoIH3wBkPANWTng56VeXiItkh_fl5Lz3QwZIpQ6ut1AMFCw/viewform?usp=sf_link).

v2e development was supported by the Swiss National Competence Center for Robotics [(NCCR Robotics)](https://nccr-robotics.ch/).

## News

See [changelog](https://github.com/SensorsINI/v2e/blob/master/CHANGELOG.md) for latest news

## Contact
Yuhuang Hu (yuhuang.hu@ini.uzh.ch)
Tobi Delbruck (tobi@ini.uzh.ch)

### Citation
If you use v2e, we appreciate a citation to the paper below. See the [v2e home page](https://sites.google.com/view/video2events/home) for futher background papers.

+ Y. Hu, S-C. Liu, and T. Delbruck. v2e: From Video Frames to Realistic DVS Events. In _2021 IEEE/CVF  Conference  on  Computer  Vision  and  Pattern Recognition Workshops (CVPRW)_, URL: https://arxiv.org/abs/2006.07722, 2021

To reproduce the experiments of the paper, please find [this repository](https://github.com/SensorsINI/v2e_exps_public).

### Advice about conversion time
We recommend running _v2e_ on a CUDA GPU or it will be very slow, particularly when using SuperSloMo upsampling. 
Even with a low-end GTX-1050, _v2e_ runs about 50-200X slower than real time 
using 10X slowdown factor and 346x260 video.

Conversion speed depends linearly on the reciprocal of the desired DVS timestamp resolution.
If you demand fine resolution of e.g. 100us, 
then expect many minutes of computing per second of source video. Running on Google colab
with GPU, it took 500s per second of 12FPS source video, because of the very high upsampling ratio
of over 800X and the 220k frames that needed to be produced for DVS modeling.

We advise using the _--stop_ option for trial run before starting a long conversion.
## Using v2e in Google CoLab
If you don't want to install, try [opening v2e in google colab](https://colab.research.google.com/drive/1czx-GJnx-UkhFVBbfoACLVZs8cYlcr_M?usp=sharing).

## Local Installation
There are 3 general steps
1. Make a _conda_ environment
2. Install _pytoch_ and other _conda_ distributed packages to this environment
3. Install the rest of the packages and _v2e_ to the conda enviroment with _pip_.

_pip_ is needed because some packages are not availble from the conda repositories. It is important to go in this order because _conda_ in general is not aware of _pip_ installs.

### Make conda environment
Install _v2e_ to a separate Python environment
such as `conda` environment:

```bash
conda create -n v2e python=3.9  # create a new environment
conda activate v2e  # activate the environment
```
### Ensure you have CUDA enabled GPU-accelerated _pytorch_
See https://pytorch.org/get-started/locally/ to generate the correct
conda install command to enable GPU-accelerated CUDA. On windows 10, for example, this tool generated the following command. _v2e_ does not use _torchaudio_ so you can leave it out.
```
conda install pytorch torchvision torchaudio cudatoolkit=11.3 -c pytorch
```

You may want to check this stackoverflow question:
https://stackoverflow.com/questions/57238344/i-have-a-gpu-and-cuda-installed-in-windows-10-but-pytorchs-torch-cuda-is-availa

### Use pip to install rest of packages and v2e
After installing pytorch to your CUDA environment, to install v2e in developer mode (so your edits to source take effect immediately), run the following command in your terminal inside the activated conda environment. The `python -m pip install -e .` command runs _setup.py_ which installs more packages and adds a script to run _v2e_ from the command line to the _conda_ enviroment python path; see https://stackoverflow.com/questions/39023758/what-does-pip-install-dot-mean:
```bash
conda activate v2e # activate your env with pytoch already installed by conda
git clone https://github.com/SensorsINI/v2e
cd v2e
python -m pip install -e . # use pip to install requirements from setup.py in user mode (so your edits to source files in v2e take effect immediately
```

+ If you want an additional Windows GUI interface, you will need to install [Gooey](https://github.com/chriskiehl/Gooey) package. This package works the best on Windows:
    ```bash
    pip install Gooey
    ```
    On Linux, `Gooey` can be hard to install.

    For a sample of conversion using the gooey GUI, see https://youtu.be/THJqRC_q2kY

### Download SuperSloMo model

We use the excellent [Super SloMo](https://people.cs.umass.edu/~hzjiang/projects/superslomo/) framework to interpolate the APS frames.
However, since APS frames only record light intensity, we  retrained it on grayscale images.

Download our pre-trained model checkpoint from Google Drive
[[SuperSloMo39.ckpt]](https://drive.google.com/u/0/uc?id=17QSN207h05S_b2ndXjLrqPbBTnYIl0Vb&export=download) (151 MB) and save it to the _input_ folder. 
The default value of --slomo_model argument is set to this location.

### Download sample input data
The sample input videos to try _v2e_ with are in [v2e-sample-input-data](https://drive.google.com/drive/folders/1oWxTB9sMPp6UylAdxg5O1ko_GaIXz7wo?usp=sharing) on google drive.

Download the [tennis.mov](https://drive.google.com/file/d/1dNUXJGlpEM51UVYH4-ZInN9pf0bHGgT_/view?usp=sharing)
video and put in the _input_ folder
to run the example below.

## Usage

_v2e_ serves multiple purposes. Please read to code if you would like to adapt it for your own application. Here, we only introduce the usage for generating DVS events from conventional video and from specific datasets.

## Render emulated DVS events from conventional video.

_v2e.py_ reads a standard video (e.g. in .avi, .mp4, .mov, or .wmv), or a folder of images, and generates emulated DVS events at upsampled timestamp resolution.

Don't be intimidated by the huge number of options. Running _v2e.py_ with no arguments sets reasonable values and opens a file browser to let you select an input video. Inspect the logging output for hints.

**Hint:** Note the options _[--dvs128 | --dvs240 | --dvs346 | --dvs640 | --dvs1024]_; they set output size and width to popular DVS cameras.

**On headless platforms**, with no graphics output, use --no_preview option to suppress the OpenCV windows.

```
usage: v2e.py [-h] [-o OUTPUT_FOLDER] [--avi_frame_rate AVI_FRAME_RATE]
              [--output_in_place [OUTPUT_IN_PLACE]] [--overwrite]
              [--unique_output_folder [UNIQUE_OUTPUT_FOLDER]]
              [--auto_timestamp_resolution [AUTO_TIMESTAMP_RESOLUTION]]
              [--timestamp_resolution TIMESTAMP_RESOLUTION]
              [--dvs_params DVS_PARAMS] [--pos_thres POS_THRES]
              [--neg_thres NEG_THRES] [--sigma_thres SIGMA_THRES]
              [--cutoff_hz CUTOFF_HZ] [--leak_rate_hz LEAK_RATE_HZ]
              [--shot_noise_rate_hz SHOT_NOISE_RATE_HZ]
              [--photoreceptor_noise]
              [--leak_jitter_fraction LEAK_JITTER_FRACTION]
              [--noise_rate_cov_decades NOISE_RATE_COV_DECADES]
              [--refractory_period REFRACTORY_PERIOD]
              [--dvs_emulator_seed DVS_EMULATOR_SEED]
              [--show_dvs_model_state SHOW_DVS_MODEL_STATE [SHOW_DVS_MODEL_STATE ...]]
              [--save_dvs_model_state] [--output_height OUTPUT_HEIGHT]
              [--output_width OUTPUT_WIDTH]
              [--dvs128 | --dvs240 | --dvs346 | --dvs640 | --dvs1024]
              [--disable_slomo] [--slomo_model SLOMO_MODEL]
              [--batch_size BATCH_SIZE] [--vid_orig VID_ORIG]
              [--vid_slomo VID_SLOMO] [--slomo_stats_plot] [-i INPUT]
              [--input_frame_rate INPUT_FRAME_RATE]
              [--input_slowmotion_factor INPUT_SLOWMOTION_FACTOR]
              [--start_time START_TIME] [--stop_time STOP_TIME] [--crop CROP]
              [--hdr] [--synthetic_input SYNTHETIC_INPUT]
              [--dvs_exposure DVS_EXPOSURE [DVS_EXPOSURE ...]]
              [--dvs_vid DVS_VID] [--dvs_vid_full_scale DVS_VID_FULL_SCALE]
              [--skip_video_output] [--no_preview] [--ddd_output]
              [--dvs_h5 DVS_H5] [--dvs_aedat2 DVS_AEDAT2]
              [--dvs_text DVS_TEXT] [--cs_lambda_pixels CS_LAMBDA_PIXELS]
              [--cs_tau_p_ms CS_TAU_P_MS] [--scidvs]
v2e: generate simulated DVS events from video.
optional arguments:
  -h, --help            show this help message and exit
Output: General:
  -o OUTPUT_FOLDER, --output_folder OUTPUT_FOLDER
                        folder to store outputs.
  --avi_frame_rate AVI_FRAME_RATE
                        frame rate of output AVI video files; only affects
                        playback rate.
  --output_in_place [OUTPUT_IN_PLACE]
                        store output files in same folder as source video (in
                        same folder as frames if using folder of frames).
  --overwrite           overwrites files in existing folder (checks existence
                        of non-empty output_folder).
  --unique_output_folder [UNIQUE_OUTPUT_FOLDER]
                        If specifying --output_folder, makes unique output
                        folder based on output_folder, e.g. output1 (if non-
                        empty output_folder already exists)
DVS timestamp resolution:
  --auto_timestamp_resolution [AUTO_TIMESTAMP_RESOLUTION]
                        (Ignored by --disable_slomo or --synthetic_input.) If
                        True (default), upsampling_factor is automatically
                        determined to limit maximum movement between frames to
                        1 pixel. If False, --timestamp_resolution sets the
                        upsampling factor for input video. Can be combined
                        with --timestamp_resolution to ensure DVS events have
                        at most some resolution.
  --timestamp_resolution TIMESTAMP_RESOLUTION
                        (Ignored by --disable_slomo or --synthetic_input.)
                        Desired DVS timestamp resolution in seconds;
                        determines slow motion upsampling factor; the video
                        will be upsampled from source fps to achieve the at
                        least this timestamp resolution.I.e. slowdown_factor =
                        (1/fps)/timestamp_resolution; using a high resolution
                        e.g. of 1ms will result in slow rendering since it
                        will force high upsampling ratio. Can be combind with
                        --auto_timestamp_resolution to limit upsampling to a
                        maximum limit value.
DVS model:
  --dvs_params DVS_PARAMS
                        Easy optional setting of parameters for DVS
                        model:None, 'clean', 'noisy'; 'clean' turns off noise,
                        sets unlimited bandwidth and makes threshold variation
                        small. 'noisy' sets limited bandwidth and adds leak
                        events and shot noise.This option by default will
                        disable user set DVS parameters. To use custom DVS
                        paramters, use None here.
  --pos_thres POS_THRES
                        threshold in log_e intensity change to trigger a
                        positive event.
  --neg_thres NEG_THRES
                        threshold in log_e intensity change to trigger a
                        negative event.
  --sigma_thres SIGMA_THRES
                        1-std deviation threshold variation in log_e intensity
                        change.
  --cutoff_hz CUTOFF_HZ
                        photoreceptor IIR lowpass filter cutoff-off 3dB
                        frequency in Hz - see
                        https://ieeexplore.ieee.org/document/4444573.CAUTION:
                        See interaction with timestamp_resolution and
                        auto_timestamp_resolution; check output logger
                        warnings. The input sample rate (frame rate) must be
                        fast enough to for accurate IIR lowpass filtering.
  --leak_rate_hz LEAK_RATE_HZ
                        leak event rate per pixel in Hz - see
                        https://ieeexplore.ieee.org/abstract/document/7962235
  --shot_noise_rate_hz SHOT_NOISE_RATE_HZ
                        Temporal noise rate of ON+OFF events in darkest parts
                        of scene; reduced in brightest parts.
  --photoreceptor_noise
                        Create temporal noise by injecting Gaussian noise to
                        the log photoreceptor before lowpass filtering.This
                        way, more accurate statistics of temporal noise will
                        tend to result but the noise rate will only
                        approximate the desired noise rate; the photoreceptor
                        noise will be computed to result in the
                        --shot_noise_rate noise value. Overrides the default
                        shot noise mechanism reported in 2020 v2e paper.
  --leak_jitter_fraction LEAK_JITTER_FRACTION
                        Jitter of leak noise events relative to the (FPN)
                        interval, drawn from normal distribution
  --noise_rate_cov_decades NOISE_RATE_COV_DECADES
                        Coefficient of Variation of noise rates (shot and
                        leak) in log normal distribution decades across pixel
                        arrayWARNING: currently only in leak events
  --refractory_period REFRACTORY_PERIOD
                        Refractory period in seconds, default is 0.5ms.The new
                        event will be ignore if the previous event is
                        triggered less than refractory_period ago.Set to 0 to
                        disable this feature.
  --dvs_emulator_seed DVS_EMULATOR_SEED
                        Set to a integer >0 to use a fixed random seed.default
                        is 0 which means the random seed is not fixed.
  --show_dvs_model_state SHOW_DVS_MODEL_STATE [SHOW_DVS_MODEL_STATE ...]
                        One or more space separated list model states. Do not
                        use '='. E.g. '--show_dvs_model_state all'. Possible
                        models states are (without quotes) either 'all' or
                        chosen from dict_keys(['new_frame', 'log_new_frame',
                        'lp_log_frame', 'scidvs_highpass',
                        'photoreceptor_noise_arr', 'cs_surround_frame',
                        'c_minus_s_frame', 'base_log_frame', 'diff_frame'])
  --save_dvs_model_state
                        save the model states that are shown (cf
                        --show_dvs_model_state) to avi files
DVS camera sizes (selecting --dvs346, --dvs640, etc. overrides --output_width and --output_height:
  --output_height OUTPUT_HEIGHT
                        Height of output DVS data in pixels. If None, same as
                        input video. Use --output_height=260 for Davis346.
  --output_width OUTPUT_WIDTH
                        Width of output DVS data in pixels. If None, same as
                        input video. Use --output_width=346 for Davis346.
  --dvs128              Set size for 128x128 DVS (DVS128)
  --dvs240              Set size for 240x180 DVS (DAVIS240)
  --dvs346              Set size for 346x260 DVS (DAVIS346)
  --dvs640              Set size for 640x480 DVS
  --dvs1024             Set size for 1024x768 DVS
SloMo upsampling (see also "DVS timestamp resolution" group):
  --disable_slomo       Disables slomo interpolation; the output DVS events
                        will have exactly the timestamp resolution of the
                        source video (which is perhaps modified by
                        --input_slowmotion_factor).
  --slomo_model SLOMO_MODEL
                        path of slomo_model checkpoint.
  --batch_size BATCH_SIZE
                        Batch size in frames for SuperSloMo. Batch size 8-16
                        is recommended if your GPU has sufficient memory.
  --vid_orig VID_ORIG   Output src video at same rate as slomo video (with
                        duplicated frames).
  --vid_slomo VID_SLOMO
                        Output slomo of src video slowed down by
                        slowdown_factor.
  --slomo_stats_plot    show a plot of slomo statistics
Input file handling:
  -i INPUT, --input INPUT
                        Input video file or a image folder; leave empty for
                        file chooser dialog.If the input is a folder, the
                        folder should contain a ordered list of image files.In
                        addition, the user has to set the frame rate manually.
  --input_frame_rate INPUT_FRAME_RATE
                        Either override the video file metadata frame rate or
                        manually define the video frame rate when the video is
                        presented as a list of image files. Overrides the
                        stored (metadata) frame rate of input video. This
                        option overrides the --input_slowmotion_factor
                        argument in case the input is from a video file.
  --input_slowmotion_factor INPUT_SLOWMOTION_FACTOR
                        (See --input_frame_rate argument too.) Sets the known slow-motion factor of the input video, 
                        i.e. how much the video is slowed down, i.e., 
                        the ratio of shooting frame rate to playback frame rate. 
                        input_slowmotion_factor<1 for sped-up video and 
                        input_slowmotion_factor>1 for slowmotion video.
                        If an input video is shot at 120fps yet is presented as a 30fps video 
                        (has specified playback frame rate of 30Hz, 
                        according to file's FPS setting), 
                        then set --input_slowdown_factor=4.
                        It means that each input frame represents (1/30)/4 s=(1/120)s.
                        If input is video with intended frame intervals of 
                        1ms that is in AVI file 
                        with default 30 FPS playback spec, 
                        then use ((1/30)s)*(1000Hz)=33.33333.
  --start_time START_TIME
                        Start at this time in seconds in video. Use None to
                        start at beginning of source video.
  --stop_time STOP_TIME
                        Stop at this time in seconds in video. Use None to end
                        at end of source video.
  --crop CROP           Crop input video by (left, right, top, bottom) pixels.
                        E.g. CROP=(100,100,0,0) crops 100 pixels from left and
                        right of input frames. CROP can also be specified as
                        L,R,T,B without ()
  --hdr                 Treat input video as high dynamic range (HDR)
                        logarithmic, i.e. skip the linlog conversion step. Use
                        --hdr for HDR input with floating point gray scale
                        input videos. Units of log input are based on white
                        255 pixels have values ln(255)=5.5441
Synthetic input:
  --synthetic_input SYNTHETIC_INPUT
                        Input from class SYNTHETIC_INPUT that has methods
                        next_frame() and total_frames(). Disables file input
                        and SuperSloMo frame interpolation and the DVS
                        timestamp resolution is set by the times returned by
                        next_frame() method. SYNTHETIC_INPUT.next_frame()
                        should return a tuple (frame, time) with frame having
                        the correct resolution (see DVS model arguments) which
                        is array[y][x] with pixel [0][0] at upper left corner
                        and pixel values 0-255. The time is a float in
                        seconds. SYNTHETIC_INPUT must be resolvable from the
                        classpath. SYNTHETIC_INPUT is the module name without
                        .py suffix. See example moving_dot.py.
Output: DVS video:
  --dvs_exposure DVS_EXPOSURE [DVS_EXPOSURE ...]
                        Mode to finish DVS frame event integration:
                        	duration time: Use fixed accumulation time in seconds, e.g. 
                        		--dvs_exposure duration .005; 
                        	count n: Count n events per frame,e.g.
                        		--dvs_exposure count 5000;
                        	area_count M N: frame ends when any area of N x N pixels fills with M events, e.g.
                        		-dvs_exposure area_count 500 64
                        	source: each DVS frame is from one source frame (slomo or original, depending on if slomo is used)
  --dvs_vid DVS_VID     Output DVS events as AVI video at frame_rate. To
                        suppress, supply argument None.
  --dvs_vid_full_scale DVS_VID_FULL_SCALE
                        Set full scale event count histogram count for DVS
                        videos to be this many ON or OFF events for full white
                        or black.
  --skip_video_output   Skip producing video outputs, including the original
                        video, SloMo video, and DVS video. This mode also
                        prevents showing preview of output (cf --no_preview).
  --no_preview          disable preview in cv2 windows for faster processing.
Output: DVS events:
  --ddd_output          Save frames, frame timestamp and corresponding event
                        index in HDF5 format used for DDD17 and DDD20
                        datasets. Default is False.
  --dvs_h5 DVS_H5       Output DVS events as hdf5 event database.
  --dvs_aedat2 DVS_AEDAT2
                        Output DVS events as DAVIS346 camera AEDAT-2.0 event
                        file for jAER; one file for real and one file for v2e
                        events. To suppress, supply argument None.
  --dvs_text DVS_TEXT   Output DVS events as text file with one event per line
                        [timestamp (float s), x, y, polarity (0,1)].
Center-Surround DVS:
  --cs_lambda_pixels CS_LAMBDA_PIXELS
                        space constant of surround in pixels, None to disable.
                        This space constant lambda is sqrt(1/gR) where g is
                        the transverse conductance and R is the lateral
                        resistance.
  --cs_tau_p_ms CS_TAU_P_MS
                        time constant of photoreceptor center of diffuser in
                        ms, or 0 to disable for instantaneous surround.
                        Defined as C/g where C is capacitance and g is the
                        transverse conductance from photoreceptor to
                        horizontal cell network. This time isthe time constant
                        for h cell diffuser network response time to global
                        input to photoreceptors. If set to zero, then the
                        simulation of diffuser runs until it converges, i.e.
                        until the maximum change between timesteps is smaller
                        than a threshold value
SCIDVS pixel:
  --scidvs              Simulate proposed SCIDVS pixel with nonlinear
                        adapatation and high gain
Run with no --input to open file dialog
```
You can put [tennis.mov](https://drive.google.com/file/d/1dNUXJGlpEM51UVYH4-ZInN9pf0bHGgT_/view?usp=sharing) in the _input_ folder to try it out with the command line below.  Or leave out all options and just use the file chooser to select the movie.

From root of v2e, run the following
```

CHANGELOG.md
# v1.6.2


## Bug fixes

 - Fix x,y output order in ae_text_output
 - Add warnings about extraneous extra arguments if there is no synthetic input class to pass them to
 - Document the order of fields in event generation in emulator.py
 - Fixed usage command to correctly show DVS output size options and to warn users about missing output size
 - Fixed docstrings for methods to correct x,y order of event fields

# v1.6.0

##  New features
  - **HDR (high dynamic range)**: The **_--hdr option_** treats inputs as logarithmic gray scale. See  [scripts/particles.py](https://github.com/SensorsINI/v2e/blob/master/scripts/particles.py) for script that that uses the --hdr argument to generate moving particles with very high brightness compared to background. At line 77, see
```python
       if self.parent_args.hdr:
            self.bg=np.log(self.bg)
            self.fg=np.log(self.fg)
```
  - **More accurate shot noise modeling**: The **-_-photoreceptor_noise_** option simulates shot noise events by injecting a 1st-order IIR lowpass (RC) filtered white noise source right after photoreceptor lowpass filtering (the noise is summed here to avoid intensity-dependent lowpass filtering of the noise). The result is more realistic statistics of shot noise events.
##    Bug fixes
  - fixed DDD conversion script [ddd/ddd_extract_data.py](https://github.com/SensorsINI/v2e/blob/master/dataset_scripts/ddd/ddd_extract_data.py#L12-L12)  to work properly.
  - improved argument usage descriptions.
  - fixed order of x,y events in AEDAT text output in [v2ecore/output/ae_text_output.py](https://github.com/SensorsINI/v2e/blob/master/dataset_scripts/ddd/ddd_extract_data.py#L12-L12).
  - fixed several synthetic input scripts to work and include option arguments
  - fixed logging to log the script extra options

# v1.5.0
##    New Features
  - v2e can display and save as AVI videos all the internal DVS model states via the **_--show_dvs_model_state_** and **_--save_dvs_model_state_** options.
  - v2e can model the **CSDVS** (center-surround DVS) proposed in Delbruck, Tobi, Chenghan Li, Rui Graca, and Brian Mcreynolds. 2022. “_Utility and Feasibility of a Center Surround Event Camera_.” arXiv [cs.CV]. arXiv. http://arxiv.org/abs/2202.13076.
##    Bug fixes
- More improvements to user experience with improved CUDA supported pytorch installation instructions and warnings printed if CUDA is not available.
- More bug fixes for synthetic input script handling and more example scripts, e.g. _spots.py_.
- Displayed DVS model state cv2 windows can now be dragged and resized by included cv2.pollkey() calls periodically.


# v1.4.3
##    New Features
  - Try [opening v2e in google colab](https://colab.research.google.com/drive/1czx-GJnx-UkhFVBbfoACLVZs8cYlcr_M?usp=sharing).
  - There are improved noise models of leak and shot noise.
    Both now model frozen log normal rate distributions; see the **_--noise_rate_cov_decades_** option.
    The leak noise events have timing jitter that matches observed jitter; see **_--leak_jitter_fraction_** option.
  - v2e models finite refractory period; see the **_--refractory_period_** option.
  - _setup.py_ includes dependencies and now installs _v2e_ as a script to your conda environment to run from command line.
  -  HDF output will also output the frames, to allow modeling of DAVIS cameras with APS+DVS output.
  - Added ability to generate input frames from python class.
    See [base_synthetic_input.py](https://github.com/SensorsINI/v2e/blob/master/v2ecore/base_synthetic_input.py) and command line argument **--synthetic_input**=<module_name>.
  - Synthetic input can recieve command line arguments; see [particle.py](https://github.com/SensorsINI/v2e/blob/master/scripts/particles.py).
  - Desktop notification is  generated for long running conversion completion.
  - Ability to crop input frames added; see --crop_input option.
##    Bug fixes
- Better argument checking and error reporting.
- For input folder, --start option skips frames without needing to load them from disk for huge speedup.


# v1.4.2
##    New Features
- The `EventEmulator` is now accelerated by PyTorch. If GPU is not available, the emulation will use CPU. Early results show that the emulation on GPU and CPU has similar performance.
- When saving with HDF5, the emulator will also optionally save three additional dataset: 1. `frames` luma frames, 2. `frame_ts` frame timestamps and 3. "frame_idxs" corresponding event index for dual modality saving and faster indexing. At default, this is disabled for faster and cheaper emulation. Can be enabled by adding `--davis_output` flag. This feature is currently only available for video, not synthetic input.
- More realistic leak events model. Add two options: `leak_jitter_fraction` and `noise_rate_cov_decades` to control leak events variation. Disable by setting them to 0.
- Add `refractory_period`: events that arrives at time smaller than the `refractory_period` will be filtered. Disable by setting it to 0.

##    Bug fixes
- A shot noise event and a regular event can occur at the same timestamp. This is fixed. Now if these two types of events occurred together, only one event will be output.

# v1.4.1
##    New Features
- a new option --dvs_emulator_seed is introduced. This option allows the user to use a fixed random seed. The default is 0 which means the emulator uses a system defined random seed.

##    Bug fixes
- Shot noise events polarity was wrongly set to 1 always. This makes the shot noise events are always positive events. Fixed this in emulator.

scripts/csdvs.sh
#!/bin/bash
# scripts to reproduce results in 2022 CSDVS ICIP paper Special session 20.11 ‘Neuromorphic and perception-based image acquisition and analysis”
# UTILITY AND FEASIBILITY OF A CENTER SURROUND EVENT CAMERA

# spots
# parame for CSDVS
lambda=10
tau=.5
dt=.0001
seq=spots
contrast=2
thr=.2
sig_thr=0.02
cutoff=100
refr=1e-3
exp="duration .01"

#v2e --output_folder=output/$seq-dvs --pos_thr=$thr --neg_thr=$thr --sigma_thr=$sig_thr --cutoff_hz=$cutoff --refractory_period=$refr --leak_rate_hz=0 --shot_noise_rate_hz=0  --vid_orig=inten --dvs_vid=dvs --unique_output --dvs_aedat2=dvs --output_width=346 --output_height=260 --batch=1 --disable_slomo --synthetic_input=scripts.spots --dt=$dt --freq=20 --contrast=$contrast --dvs_exposure $exp --shot_noise_rate_hz=1 --show_dvs_model_state all --save_dvs_model_state
#v2e --output_folder=output/$seq-csdvs-l$lambda-t$tau --cs_lambda_pixels=$lambda --cs_tau_p_ms=$tau  --pos_thr=$thr --neg_thr=$thr --sigma_thr=$sig_thr --cutoff_hz=$cutoff --refractory_period=$refr --leak_rate_hz=0 --shot_noise_rate_hz=0  --vid_orig=inten --dvs_vid=csdvs --unique_output --dvs_aedat2=csdvs --output_width=346 --output_height=260 --batch=1 --disable_slomo --synthetic_input=scripts.spots --dt=$dt --freq=20 --contrast=$contrast --dvs_exposure $exp --shot_noise_rate_hz=1 --show_dvs_model_state all --save_dvs_model_state

# gradients
seq=gradients

#v2e  --output_folder output/$seq-dvs --pos_thr=$thr --neg_thr=$thr --sigma_thr=$sig_thr --cutoff_hz=$cutoff --refractory_period=$refr --leak_rate_hz=0 --shot_noise_rate_hz=0  --vid_orig=inten --dvs_vid=dvs --unique_output --dvs_aedat2=dvs --output_width=346 --output_height=260 --batch=1 --disable_slomo --dt=$dt --contrast=$contrast --dvs_exposure $exp --shot_noise_rate_hz=1 --show_dvs_model_state all --save_dvs_model_state --synthetic_input=scripts.gradients
#v2e  --output_folder output/$seq-csdvs  --cs_lambda_pixels=$lambda --cs_tau_p_ms=$tau --pos_thr=$thr --neg_thr=$thr --sigma_thr=$sig_thr --cutoff_hz=$cutoff --refractory_period=$refr --leak_rate_hz=0 --shot_noise_rate_hz=0  --vid_orig=inten --dvs_vid=dvs --unique_output --dvs_aedat2=dvs --output_width=346 --output_height=260 --batch=1 --disable_slomo --dt=$dt --contrast=$contrast --dvs_exposure $exp --shot_noise_rate_hz=1 --show_dvs_model_state all --save_dvs_model_state --synthetic_input=scripts.gradients

# cloudy sky
seq=cloudy-sky
input=input/CloudySky_Slomo.mp4
lambda=10
tau=2
slomo=33

#v2e --output_folder=output/$seq-dvs --input=$input --input_slowmotion_factor=$slomo --disable_slomo  --pos_thr=$thr --neg_thr=$thr --sigma_thr=$sig_thr --cutoff_hz=$cutoff --refractory_period=$refr --leak_rate_hz=0 --vid_orig=inten --dvs_vid=dvs --unique_output --dvs_aedat2=dvs --dvs346  --dvs_exposure=source
#v2e --cs_lambda_pixels=$lambda --cs_tau_p_ms=$tau --output_folder=output/$seq-csdvs --input=$input --input_slowmotion_factor=$slomo --disable_slomo  --pos_thr=$thr --neg_thr=$thr --sigma_thr=$sig_thr --cutoff_hz=$cutoff --refractory_period=$refr --cutoff_hz=100 --leak_rate_hz=0 --vid_orig=inten --dvs_vid=csdvs --unique_output --dvs_aedat2=csdvs --dvs346  --dvs_exposure=source

# flicker
seq=flicker
start=2
stop=12
input=input/flicker.mp4
lambda=10
tau=.5
slomo=33
dt=3e-4
thr=.2
sig_thr=0.02
cutoff=100
refr=1e-3

#v2e --timestamp_resolution=$dt --auto_timestamp=False --output_folder=output/$seq-dvs --start=$start --stop=$stop --input_slowmotion_factor=$slomo --input=$input --vid_orig=$seq-input   --pos_thr=$thr --neg_thr=$thr --sigma_thr=$sig_thr --cutoff_hz=$cutoff --refractory_period=$refr --leak_rate_hz=0 --dvs_vid=dvs --unique_output --dvs_aedat2=dvs --dvs346  --dvs_exposure source --cutoff_hz=$cutoff
#v2e --cs_lambda_pixels=$lambda --cs_tau_p_ms=$tau --timestamp_resolution=$dt --auto_timestamp=False  --output_folder=output/$seq-csdvs --start=$start --stop=$stop --input_slowmotion_factor=$slomo --input=$input  --vid_orig=$seq-input ---pos_thr=$thr --neg_thr=$thr --sigma_thr=$sig_thr --cutoff_hz=$cutoff --refractory_period=$refr --leak_rate_hz=0 --dvs_vid=csdvs --unique_output --dvs_aedat2=dvs --dvs346 --dvs_exposure=source --cutoff_hz=$cutoff


# csdvs-flashing-leds
seq=shadows
start=2
stop=7
input=input/csdvs/shadows/20220507_110937.mp4
lambda=4
tau=.1
slomo=32 # 960 FPS/30FPS=32
thr=0.05
cutoff=300
refr=.1e-3
#
#v2e --timestamp_resolution=.001 --auto_timestamp=False --output_folder=output/$seq-dvs --start=$start --stop=$stop --input_slowmotion_factor=$slomo --input=$input --vid_orig=$seq-input   --pos_thr=$thr --neg_thr=$thr --sigma_thr=$sig_thr --cutoff_hz=$cutoff --refractory_period=$refr --cutoff_hz=$cutoff --refractory_period=$refr --leak_rate_hz=0 --dvs_vid=dvs --unique_output --dvs_aedat2=dvs --dvs346
#v2e --cs_lambda_pixels=$lambda --cs_tau_p_ms=$tau --disable_slomo --output_folder=output/$seq-csdvs --start=$start --stop=$stop --input_slowmotion_factor=$slomo --input=$input  --vid_orig=$seq-input ---pos_thr=$thr --neg_thr=$thr --sigma_thr=$sig_thr --cutoff_hz=$cutoff --refractory_period=$refr --cutoff_hz=$cutoff --refractory_period=$refr --leak_rate_hz=0 --dvs_vid=csdvs --unique_output --dvs_aedat2=csdvs --dvs346 --dvs_exposure=source  --show_dvs_model_state all --save_dvs_model_state


